File: hooks/post-edit-track.py
#!/usr/bin/env python3
# SPDX-FileCopyrightText: 2026 Igor Santos
# SPDX-License-Identifier: MIT
# PostToolUse hook on Edit/Write/NotebookEdit: record edited absolute path + ts
# to per-session edits.jsonl. Consumed by preread-edit-check.py + statusline.

import json
import os
import sys
import time

sys.path.insert(0, os.path.join(os.path.dirname(os.path.abspath(__file__)), "lib"))
import common as c  # noqa: E402


def main() -> int:
    dir_ = c.session_dir()
    if not dir_:
        return 0

    tool = c.field(".tool_name")
    if tool not in ("Edit", "Write", "NotebookEdit"):
        return 0

    # Different tools use different field names; try both common ones.
    path = c.field(".tool_input.file_path")
    if not path:
        path = c.field(".tool_input.notebook_path")
    if not path:
        return 0

    abs_path = c.abspath(path)
    ts = int(time.time())

    line = json.dumps({"path": abs_path, "ts": ts}, separators=(",", ":"), ensure_ascii=False)
    c.atomic_append(os.path.join(dir_, "edits.jsonl"), line)

    # Bump human-readable edit count (used by statusline).
    c.incr_counter(os.path.join(dir_, "edit-count"))
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: src/hooks/post_edit_track.rs
// SPDX-FileCopyrightText: 2026 Igor Santos
// SPDX-License-Identifier: MIT

//! Ports hooks/post-edit-track.py: a PostToolUse hook on Edit/Write/
//! NotebookEdit that records the edited absolute path plus a timestamp to
//! the per-session edits.jsonl file. Consumed by preread-edit-check.rs and
//! the statusline. Emits nothing on stdout; that silence is the contract.

use crate::common::payload::Payload;
use crate::common::{abspath, atomic_append, incr_counter, session_dir};
use serde::Serialize;
use std::time::{SystemTime, UNIX_EPOCH};

#[derive(Serialize)]
struct EditRecord<'a> {
    path: &'a str,
    ts: i64,
}

pub fn run(payload: &Payload) {
    let dir = session_dir(payload);
    if dir.is_empty() {
        return;
    }

    let tool = payload.field(".tool_name");
    if tool != "Edit" && tool != "Write" && tool != "NotebookEdit" {
        return;
    }

    // Different tools use different field names; try both common ones.
    let mut path = payload.field(".tool_input.file_path");
    if path.is_empty() {
        path = payload.field(".tool_input.notebook_path");
    }
    if path.is_empty() {
        return;
    }

    let abs_path = abspath(&path);
    let record = EditRecord {
        path: &abs_path,
        ts: now_unix_seconds(),
    };
    if let Ok(line) = serde_json::to_string(&record) {
        atomic_append(&format!("{dir}/edits.jsonl"), &line);
    }

    // Bump human-readable edit count (used by statusline).
    incr_counter(&format!("{dir}/edit-count"));
}

/// Current unix time in whole seconds, matching python's `int(time.time())`
/// truncation. Never panics; a clock reading before the epoch falls back to
/// 0 rather than erroring.
fn now_unix_seconds() -> i64 {
    SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .map(|d| d.as_secs() as i64)
        .unwrap_or(0)
}
